### PR TITLE
Logging with multiprocessing

### LINKED ISSUE #1418: Log files and multiprocessing
Our current logging infrastructure is not compatible with running multiple processes via Python's `multiprocessing` module. I guess working on examples that use `multiprocessing` prompted @schmitts' question about switching off the log file (#1407)?
A minor issue is that the log file is not very useful with multiple processes, since all processes write in no specific order. A bigger issue is that our log file rotation will raise an error if the file becomes big enough to be rotated – several processes try to move/delete the file, leading to a `FileNotFoundError`. I don't think there's a need to deal with this in a perfect way (e.g. by writing a multiprocess-safe logging handler, moving the log rotation into a single process, or similar), but we should at least mention this issue in the documentation. Or, given that few users actually find such details in the documentation, we could raise a warning in the logging code when 1) a file log is used and 2) it is accessed from multiple processes. A downside of this warning would be that I don't see an easy way to only raise it once, instead of once for every process launched by `multiprocessing`. But I guess this is ok, since we really want the user to do something about it.
There are three workarounds that we can recommend to users:
1. Use a preference file to switch off file logging
2. Set `BrianLogger.tmp_log` to `None` to switch off file logging in an ad-hoc way
3. Call `BrianLogger.initialize()` at the beginning of each spawned process, which will make it use its own log file

### PR DESCRIPTION
**Pull request by mstimberg**
See the discussion in #1418. This PR changes the default file log level to DEBUG, switches off file logging when the logger is accessed from a process different from the one that called `BrianLogger.initialize()`, and updated the documentation accordingly.

@schmitts & @thesamovar, let me know if anything is unclear or could be improved.

Fixes #1418

### CODE REVIEW

**Comment by mstimberg** [comment]
There were a few issues with running the multiprocessing tests under OS X and Windows (of course there were), but now all tests pass on all platforms.

**Comment by mstimberg** [comment]
Hi @bdevans, @thesamovar any comments on the PR? I think it is rather safe to merge in the sense that it should clearly improve things, but if you could have a quick look I'd be grateful. I'd prefer to rather merge it soon, so I propose to turn this into a "merge unless someone objects" instead of waiting for a positive review :blush:

**Comment by thesamovar** [comment]
Have you tried running actual examples (not tests) in Windows? I feel like with multiprocessing stuff that's important because there are all sorts of weirdnesses and I don't know how well the test suite reproduces that?

**Comment by mstimberg** [comment]
I did run the [multiprocessing examples](https://brian2.readthedocs.io/en/stable/examples/index.html#multiprocessing) and [Graupner & Brunel (2012)](https://brian2.readthedocs.io/en/latest/examples/frompapers.Graupner_Brunel_2012.html). They ran successfully – I did not verify in detail whether it actually switches off the log, though. I guess if the corresponding test passes, there is no reason this should not work?
The [Maass et al. (2002)](https://brian2.readthedocs.io/en/latest/examples/frompapers.Maass_Natschlaeger_Markram_2002.html) example also uses multiprocessing, but I recently realized it does not work under Windows (a function should be declared on the module level instead of in the `__main__` block, I think). But it does not work in `master`, either, so this problem is unrelated to this PR.

File: docs_sphinx/advanced/logging.rst
Comment (bdevans) [inline docs_sphinx/advanced/logging.rst:41]: Typo in parallel

File: dev/continuous-integration/run_test_suite.py
Comment (bdevans) [inline dev/continuous-integration/run_test_suite.py:70]: There's a cheeky extra space here!

Comment (bdevans) [inline dev/continuous-integration/run_test_suite.py:70]: There's a cheeky extra space here!

File: docs_sphinx/advanced/logging.rst
Comment (bdevans) [inline docs_sphinx/advanced/logging.rst:46]: "wish to use" or "consider using" would be slightly more natural here.

Comment (bdevans) [inline docs_sphinx/advanced/logging.rst:46]: "wish to use" or "consider using" would be slightly more natural here.

**Review by bdevans — CHANGES_REQUESTED** [review]
A few minor tweaks to make but otherwise looks fine. I might lean towards generating a separate log file for each process, calling them something like `f"{log_file_basename}_{pid}.log"` but that might generate too much output so having just one log file for the master process is fine.

**Comment by mstimberg** [comment]
Thanks @bdevans for the review, I think the issues should be fixed now.

> I might lean towards generating a separate log file for each process, calling them something like `f"{log_file_basename}_{pid}.log"` but that might generate too much output so having just one log file for the master process is fine.

Yes, I agree that this would make sense as well. On the other hand, the logged information in the subprocesses will most likely be very similar (if not identical), and if you launch many of them, this could become annoying. IMO, the vast majority of users do not care about the log, and for those who do, the manual method gives them a way to have individual logs per subprocess.

**Comment by bdevans** [comment]
Thanks Marcel, it looks good to me now but for some reason I can't approve the changes (I can only seem to dismiss or re-request the review) - any ideas what's gone wrong?

**Comment by thesamovar** [comment]
> I did run the [multiprocessing examples](https://brian2.readthedocs.io/en/stable/examples/index.html#multiprocessing) and [Graupner & Brunel (2012)](https://brian2.readthedocs.io/en/latest/examples/frompapers.Graupner_Brunel_2012.html). They ran successfully – I did not verify in detail whether it actually switches off the log, though. I guess if the corresponding test passes, there is no reason this should not work? The [Maass et al. (2002)](https://brian2.readthedocs.io/en/latest/examples/frompapers.Maass_Natschlaeger_Markram_2002.html) example also uses multiprocessing, but I recently realized it does not work under Windows (a function should be declared on the module level instead of in the `__main__` block, I think). But it does not work in `master`, either, so this problem is unrelated to this PR.

As long as it's not breaking I guess it's OK! If Ben is happy with the code I'm happy too.

**Review by bdevans — APPROVED** [review]

**Pull request merged**
